src/cmap/message_stream.ts

```diff
@@ -0,0 +1,220 @@
+import { Duplex, type DuplexOptions } from 'stream';
+
+import type { BSONSerializeOptions, Document } from '../bson';
+import { MongoDecompressionError, MongoParseError } from '../error';
+import type { ClientSession } from '../sessions';
+import { BufferPool, type Callback } from '../utils';
+import {
+  type MessageHeader,
+  OpCompressedRequest,
+  OpMsgResponse,
+  OpQueryResponse,
+  type WriteProtocolMessageType
+} from './commands';
+import { compress, Compressor, type CompressorName, decompress } from './wire_protocol/compression';
+import { OP_COMPRESSED, OP_MSG } from './wire_protocol/constants';
+
+const MESSAGE_HEADER_SIZE = 16;
+const COMPRESSION_DETAILS_SIZE = 9; // originalOpcode + uncompressedSize, compressorID
+
+const kDefaultMaxBsonMessageSize = 1024 * 1024 * 16 * 4;
+/** @internal */
+const kBuffer = Symbol('buffer');
+
+/** @internal */
+export interface MessageStreamOptions extends DuplexOptions {
+  maxBsonMessageSize?: number;
+}
+
+/** @internal */
+export interface OperationDescription extends BSONSerializeOptions {
+  started: number;
+  cb: Callback<Document>;
+  documentsReturnedIn?: string;
+  noResponse: boolean;
+  raw: boolean;
+  requestId: number;
+  session?: ClientSession;
+  agreedCompressor?: CompressorName;
+  zlibCompressionLevel?: number;
+  $clusterTime?: Document;
+}
+
+/**
+ * A duplex stream that is capable of reading and writing raw wire protocol messages, with
+ * support for optional compression
+ * @internal
+ */
+export class MessageStream extends Duplex {
```

Comment:
I think MessageStream should stay deleted 😅 (rebase issue?)

Reply:
I imagine so.. 😅